### ISSUE TITLE
Flexible Consumer Prioritization API Proposal

### ISSUE THREAD

**Issue by ggarber** [versatica/mediasoup#662]
## Feature Request

This is a proposal for a more flexible way to define the priority of the streams being consumed.

### Use case

* In a video room one of the users is presenting and producing streams from his screen and his camera.   The requirements are:
  *  Those streams from the presenting user are not disabled (consuming 0 layers) when bandwidth is limited unless all the rest of the videos have been disabled before.  
  *  The video streams from other participants are not disabled unless the consumer from the presenting streams are already receiving the lowest spatial&temporal layers (consuming 1 single layer).

### Current Behaviour

The current approach includes two levers in the Consumers:
* `priority` defines the order of consumers to get layers and the number of layers to get before passing to the next consumer
* `preferredLayers` defines the max number of layers (spatial and temporal) that the consumer will use

and the algorithm to assign the available bandwidth to all the consumers of a transport works like this in pseudocode:
```
sortedConsumers = sortConsumersByPriority(consumers)
availableBandwidth = bandwidthEstimation()
while (availableBandwidth > 0) {
  foreach(consumer in sortedConsumers) {
     while (consumer.layers < consumer.preferredLayers && layers < consumer.priority) {
        consumer.UseLayer(consumer.layers + 1)
        availableBandwidth -= consumer.usedBitrate()
     }
  }
}
```

The limitation of this approach is that we have a single setting (the `priority`) to control both the order of consumers and the number of layers to be assigned.

So if we have 3 consumers (a, b, c) and we want them to get layers in order (c, b, a) we need to set their priorities as a=1, b=2, c=3 and that means that in some cases c will get 3 layers and a and b 0 even if I would prefer to receive 1 layer for each video.

### Proposal

Change the meaning of the `priority` field in the consumer to ONLY affect the order in which consumers are assigned layers but not how many they get in each pass.

The algorithm would be just like:
```
sortedConsumers = sortConsumersByPriority(consumers)
availableBandwidth = bandwidthEstimation()
while (availableBandwidth > 0) {
  foreach(consumer in sortedConsumers) {
     if (consumer.layers < consumer.preferredLayers) {
        consumer.UseLayer(consumer.layers + 1)
        availableBandwidth -= consumer.usedBitrate()
     }
  }
}
```

#### Additional bonus point

Add one more setting `baseLayers` to mark how many layers the consumer will get before starting to assign layers to the next consumer.

For example if baseLayers = {spatial: 1, temporal: 0} for a consumer with high priority then that consumer will get the highest resolution before going to assign layers to the next consumer in order of priority.   Default would be baseLayers = { 0, 0 }

The algorithm would be something like:
```
sortedConsumers = sortConsumersByPriority(consumers)
availableBandwidth = bandwidthEstimation()
while (availableBandwidth > 0) {
  foreach(consumer in sortedConsumers) {
     while (consumer.layers < consumer.preferredLayers && consumer.layers < consumer.baseLayers) {
        consumer.UseLayer(comsumer.layers + 1)
        availableBandwidth -= consumer.usedBitrate()
     }
  }
}
```

Maybe there are better options for naming if we don't need to maintain backwards compatiblity (f.e I think preferredLayers could be called maxLayers) but that's probably out of scope.

**Comment by vpalmisano** [comment]
> So if we have 3 consumers (a, b, c) and we want them to get layers in order (c, b, a) we need to set their priorities as a=1, b=2, c=3 and that means that in some cases c will get 3 layers and a and b 0 even if I would prefer to receive 1 layer for each video.

If we prefer to receive 1 layer for each video we would set preferredLayers=1, so the condition `consumer.layers < consumer.preferredLayers && layers < consumer.priority` will fail for consumer `c` and it will never get 3 layers.

**Comment by ggarber** [comment]
> > So if we have 3 consumers (a, b, c) and we want them to get layers in order (c, b, a) we need to set their priorities as a=1, b=2, c=3 and that means that in some cases c will get 3 layers and a and b 0 even if I would prefer to receive 1 layer for each video.
> 
> If we prefer to receive 1 layer for each video we would set preferredLayers=1, so the condition `consumer.layers < consumer.preferredLayers && layers < consumer.priority` will fail for consumer `c` and it will never get 3 layers.

The preference is to receive ALL the layers for each video so we cannot use preferredLayers=1 as far as I know.
Only when the bandwidth is limited we want to receive less layers and only when everybody is receiving only 1 layer then start disabling videos.

**Comment by vpalmisano** [comment]
Example:
- 3 consumers:
  - 2 participants (a, b) with layers [500, 300, 100, 75, 50, 25] Kbps (2 spatial x 3 temporal layers)
  - 1 screenshare (c) with layers: [500, 300, 100] Kbps (3 temporal layers)
  - priorities: (a, b, c) = [1, 2, 3] 
 
1. avail. bandwidth: 10 MBps => current layers: [6, 6, 3] = [500, 500, 500] Kbps
2. avail. bandwidth: 1 Mbps => [4, 5, 3] = [100, 300, 500]

**Comment by vpalmisano** [comment]
Another solution that avoids assigning all of the available bandwidth to consumers with priority > 1:

```
sortedConsumers = sortConsumersByPriority(consumers)
availableBandwidth = bandwidthEstimation()
baseAllocation = true
while (availableBandwidth > 0) {
  foreach(consumer in sortedConsumers) {
     if (baseAllocation) {
       // try to increase by 1 layer
       usedBitrate = consumer->IncreaseLayer(availableBitrate)
       if (usedBitrate == 0) {
          // no more bandwidth for the remaining consumers
          break
       }
       availableBandwidth -= usedBitrate
     } else {
       for (i = 1; i <= consumer.priority; ++i) {
         // try to increase by 1 additional layer for each priority point
         usedBitrate = consumer->IncreaseLayer(availableBitrate)
         if (usedBitrate == 0) {
            // no more bandwidth for the remaining consumers
            break
         }
         availableBandwidth -= usedBitrate
       }
     }
  }
  baseAllocation = false
}
```

**Comment by ibc** [comment]
I like the latest proposal.

**Comment by ggarber** [comment]
ok, it doesn't allow to change the order of consumers without assigning 2x layers to some of them but I can live with it.

In the use case I mentioned:
- presenter screensharing priority=2
- presenter camera priority=2
- rest of streams priority=1

you will have specific bandwidth estimations where you will end up receiving:
3 layers, 3 layers, 1 layers from those 3 streams (or 6 layers, 6 layers, 2 layers) when you could have 2, 2, 2 layers or something more balanced.

But it is better than the current experience for sure, so I can go ahead and implement that.  Thank you

**Comment by ggarber** [comment]
Probably it makes sense to add the "flexible" approach at some time but th "simple" approach has been already implemented and landed so I'm closing this issue.  Thank you everybody!
